README
Procedural Texgen v 0.006

Overview:
    A library and a few utilities to make procedural textures.

Changelog.txt
v0.006:
    Added TARGA support
        Pros:
            You can shove your textures in a file now.
        Cons:
            Some features not totally implemented
    
    Can pass a dictionary as input to a texture
        Pros:
            You can do it!
    
    Rendering more algebraically correct
        Pros:
            More correct
        Cons:
            Feels backwards if you're dyslexic
    
    Added iterY and iterX arguments to Texture.render()
        Pros:
            Allows fuller user control over what threads render what samples
        What this means:
            You can use iterX and iterY to control what iterators are used to calculate which samples each thread is trying to
                render. This means that if you're completely braindead you can have multiple threads render the same sample.
    
    Stopped using iterators for sample positions
        Pros:
            Better precison
            Smaller code
        What this means:
            I used to use a count iterator to find the next sample's Y position, this would result in numbers like 0.999...8
                when it should be 1.00, this problem was fixed by using multiplication instead of count iterators
    
    Added thread and altch arguments to Texture.render(), added Texture.getBuffer()
        Pros:
            Allows the user to take advantage of multiple cores on a machine
        What this means:
            In order to render textures for multiple cores, the multiprocessing module must be called from the top level
                script, which means that this module can't do anything related to multiple cores. However, to make it easier
                to do this you can use the thread argument to provide a tuple/list containing the current thread and total threads
                (ex: (0, 4)), and altch to provide a dictionary of references to a multiprocessing.Array of unsigned bytes
                ex:
                    import copy, multiprocessing
                    ba = bytearray.fromhex("00000000")
                    Texture.render(Random, crap, here, doo, dee, doo, threads=(0, 4),
                                   altch={"R":multiprocessing.Array("B", ba), "G":multiprocessing.Array("B", ba),
                                        "B":multiprocessing.Array("B", ba)})
                Note that the dictionary shouldn't be a reference, the arrays should, and that the arrays must already be full
                    (since they can't be resized).
            Since the color channels aren't stored in the class, you have to access the alternate channels to get it.
            YOU ALSO CANNOT CHANGE THE SIZE/DIMENSIONS BETWEEN PROCESSES I don't know what kind of imbecile would try, but
                whatever, it's there.
            There will be problems on Windows: since it doesn't have os.fork() this means the processes can't run a method and
                since Texture.render() is a method, problems arise. Long story short Windows is like the Budweiser of OSes, it's
                fine as long as you don't expect anything from it.
    
    Notable Notes:
        Dox for some stuff
        Cleaned up code quite a bit - made it shorter, more concise
        Slightly more PEP 8 compliant
        Added a few docstrings
        Was horribly idiotic with making optional sizes/dimensions through Texture.render(), it should always work now and be as
            accurate as using default values
        Lots of small stuff

v0.0051:
    Dimensions and sizes can be called through Texture.render()
        Pros:
            The same texture can be rendered at whatever size/dimensions you feel at the moment
        What this means:
            You don't have to rebuild the texture to render a different part of it and/or at a different size
            It's entirely optional, so you still have to initialize the texture with default size/dimensions
    Notable Notes:
        Last version (0.005) had a showstopper that I should have caught, but didn't for some reason - it has been fixed
    
v0.005:
    Changed the way strokes are called
        Pros:
            Allows any stroke to run _ANY_ other (unless you like infinite loops) stroke no matter where it is
            Should be a little more efficient
            ?????
            Profit
        What this means:
            Each stroke can have a key in their settings that points to a different stroke that is called
            You can supply a color dictionary to input instead of having to call a Number/etc. stroke to set a background color
            Brushes can have multiple inputs if they so choose
    
    Put colorspace in info (where it should be)
        Pros:
            Calling strokes is easier and less complicated
        What this means:
            Colorspace used to be an argument passed to each stroke - by moving it into the info class, there are less arguments
                that need to be passed
            Making strokeA call strokeB with a different colorspace than the one passed to strokeA is now more logical
    
    Added a clearBuffer() function to the texture class
        Pros:
            Can clear the channels attribute (so it isn't full after a render() call)
        What this means:
            If a render() was called, it would store the result in the channels attribute for no good reason, now you can clear
                it manually
    
    Strokes (should) have a man function and a _cache attribute
        Pros:
            man is for documeting what settings are needed and/or used by the stroke
            _cache is for the stroke to keep data (useful during init, over iterations, etc.) AKA only the stroke should ever
                see it, and you sure as hell should never change it
        What this means:
            It's easier to provide a description of how to supply settings
            Strokes now have a standard way to store information
    
    Started documenting changes

Test.py
import Base
import itertools, math, multiprocessing, copy
import Values as V
import Brushes as B
import TARGA
from multiprocessing import Process
Cores = 2

Size = 10
print("Going over some basics:\nInit:")
TexA = Base.Texture(Size, Size, ("-1.00", "1.00"), ("-1.00", "1.00"))
TexA.strokes = (V.Number({"value":0.5}),)
TexA.input = B.Circle({"outsidecolor":TexA.strokes[0], "radius":.5, "x":0.00, "y":0.00, "insidecolor":{"R":1.00, "G":0.00, "B":0.00, "outsidecolor":TexA.strokes[0]}})
print("Init finished, rendering:")
Picture = TexA.render(colorspace="RGB")
spititout = input("Rendering finished, print?")
if spititout.lower() == "y" or spititout.lower() == "ye" or spititout.lower() == "yes" or spititout.lower() == "yep" or spititout.lower() == "t" or spititout.lower() == "true":
    for y in range(int(len(Picture["R"])/Size)):
        line = []
        for x in range(Size):
            line.append(str(Picture["R"][x+(y*Size)]).zfill(3)+"|"+str(Picture["G"][x+(y*Size)]).zfill(3)+"|"+str(Picture["B"][x+(y*Size)]).zfill(3))
        print(line)
TexA.clearBuffer()
print("TARGA file:")
w = TARGA.makeTARGA(sizeX=Size, sizeY=Size, raster=Picture)
file = open("Out1.tga", "wb")
file.write(w)
file.close()
del Picture, spititout

print("\n\nGoing over advanced stuff (like rendering upside-down)\nInit: ")
Base.LINEINFO = True
BigSize = Size*2
BigBPC = 8
Bytes = math.ceil(BigBPC/8)
TexB = Base.Texture(Size, Size, ("1.00", "-1.00"), ("1.00", "-1.00"))#Note the dimensions are reversed
Outsidestrokes = (B.Circle({"insidecolor":{"R":.5, "G":0.0, "B":1.00}, "outsidecolor":{"R":0.00, "G":0.00, "B":0.25}, "radius":.5, "x":.3, "y":.3}), )
TexB.input = B.Circle({"radius":.5, "x":0.0, "y":0.0, "insidecolor":{"R":1.00, "G":0.00, "B":0.00}, "outsidecolor":Outsidestrokes[0]})
print(TexB.input.man())
print("Init finished, rendering:")
Picture = TexB.render(sizeX=BigSize, sizeY=BigSize, dimensionX=("2.00", "-2.00"), dimensionY=("2.00", "-2.00"), colorspace="RGB", bpc=BigBPC)#Still reversedTexB.clearBuffer()
spititout = input("Rendering finished, print?")
if spititout.lower() == "y" or spititout.lower() == "ye" or spititout.lower() == "yes" or spititout.lower() == "yep" or spititout.lower() == "t" or spititout.lower() == "true":
    for y in range(int(len(Picture["R"])/(BigSize))):
        line = []
        for x in range(BigSize):
            index = (x+(y*BigSize))*Bytes
            line.append(str(Picture["R"][index]).zfill(3)+"|"+str(Picture["G"][index]).zfill(3)+"|"+str(Picture["B"][index]).zfill(3))
        print(line)
del Picture, spititout

print("\n\nGoing over some multicore stuff.\nThis shit won't work on Windows (or at all).")
Base.lineinfo = True
BigSize = Size*2
BigBPC = 8
Bytes = math.ceil(BigBPC/8)
TexC = Base.Texture(Size, Size, ("1.00", "-1.00"), ("1.00", "-1.00"))#Note the dimensions are reversed
Outsidestrokes = (B.Circle({"insidecolor":{"R":.5, "G":0.0, "B":1.00}, "outsidecolor":{"R":0.00, "G":0.00, "B":0.25}, "radius":.5, "x":.3, "y":.3}), )
TexC.input = B.Circle({"input":Outsidestrokes[0], "radius":.5, "x":0.0, "y":0.0, "insidecolor":{"R":1.00, "G":0.00, "B":0.00}})
print("Init finished")
t = bytearray.fromhex("00"*(BigSize**2))
d = {}
for s in ("R", "G", "B"):
    d[s] = multiprocessing.Array("B", copy.copy(t))

if str(input("Do multicore? Guarunteed to not work on Windows (if at all):"))[0].lower() == "y":
    proc = []
    print("Starting making processes!")
    for i in range(Cores):
        if __name__ == "__main__":
            print("Going through __main__!")
            proc.append(None)
            proc[i-1] = Process(target=TexC.render, kwargs={"colorspace":"RGB", "bpc":BigBPC, "sizeX":BigSize, "sizeY":BigSize, "dimensionX":("2.00", "-2.00"), "dimensionY":("2.00", "-2.00"), "thread":(i, Cores), "altch":copy.copy(d)})
    for i in range(Cores):
        proc[i].start()
        proc[i].join()
    print(proc)
    #for i in range(Cores):
        #TexB.render("RGB", bpc=BigBPC, sizeX=BigSize, sizeY=BigSize, dimensionX=("2.00", "-2.00"), dimensionY=("2.00", "-2.00"), thread=(i, 2), altch=copy.copy(d))#Still reversedTexB.clearBuffer()

    spititout = input("Rendering finished, print?")
    if spititout.lower() == "y" or spititout.lower() == "ye" or spititout.lower() == "yes" or spititout.lower() == "yep" or spititout.lower() == "t" or spititout.lower() == "true":
        for y in range(int(len(d["R"])/(BigSize))):
            line = []
            for x in range(BigSize):
                index = (x+(y*BigSize))*Bytes
                line.append(str(d["R"][index]).zfill(3)+"|"+str(d["G"][index]).zfill(3)+"|"+str(d["B"][index]).zfill(3))
            print(line)
    del d


Modes.py
#Python 3.2.x only, mofos
def _add(a, b, ):
    return a+b

def _multiply(a, b):
    return a*b

def _power(a, b):
    return a**b

def _rpower(a, b):
    return b**a

def _subtract(a, b):
    return a-b

def _rsubtract(a, b):
    return b-a

def _divide(a, b):
    return a/b

def _rdivide(a, b):
    return b/a

def _max(a, b):
    if a > b:
        return a
    else:
        return b

def _min(a, b):
    if a < b:
        return a
    else:
        return b

Modes = {"+": _add, "add": _add, "*": _multiply, "multiply": _multiply, "**": _power, "r**": _rpower, "power": _power,
         "rpower": _power,  "-": _subtract, "r-": _rsubtract, "subtract": _subtract, "rsubtract": _rsubtract,
         "/": _divide, "r/": _rdivide, "divide": _divide, "rdivide": _rdivide, "^": _max, "max": _max, "v": _min,
         "min": _min}

TARGA.py
#Makes a .tga file out of a raster dict
END = "big"

def makeTARGA(sizeX, sizeY, raster, ID=None, originX=0, originY=0, imagetype=None, bpp=8, developer=None, extension=None, out=False, defaultA="FF", defaultR="00", defaultG="00", defaultB="00"):
    if ID is not None:
        if len(ID) > 255: raise ValueError("ID can't be more than 255 characters long, yo")
    if imagetype is not None:
        if imagetype > 255 or (imagetype > 10 and imagetype < 128): raise ValueError("colormaptype has to be either 0, 1 or between 128 and 255")
    if len(raster) > 4: raise ValueError("Too many channels")
    if sizeX > 65535 or sizeY > 65535: raise ValueError("Image too big")
    if originX > 65535 or originY > 65535: raise ValueError("Offset too large")
    #ID length
    header = bytearray(int(0).to_bytes(1, END) if ID is None else len(ID).to_bytes(1, END))
    #Color map type
    header += int(0).to_bytes(1, END)
    #Image type
    if imagetype is None:
        if len(raster) == 0:
            header += int(0).to_bytes(1, END)
        elif len(raster) == 1:
            header += int(3).to_bytes(1, END)
        elif len(raster) >= 2 and len(raster) <= 4:
            header += int(2).to_bytes(1, END)
    else:
        header += imagetype.to_bytes(1, END)
    #Color map specification
    header += int(0).to_bytes(5, END)
    #Image specification
    header += int(originX).to_bytes(2, END)#X-Origin
    header += int(originY).to_bytes(2, END)#Y-Origin
    header += int(sizeX).to_bytes(2, "little")#Width
    header += int(sizeY).to_bytes(2, "little")#Height
    header += int(bpp*4).to_bytes(1, END)#Pixel depth
    header += bpp.to_bytes(1, END)#Alpha depth
    #Image Data
    if ID is not None: header.append(ID)#ID
    keylist = list(raster.keys())
    color = []
    for s in "BGRA":
        color.append(raster[s] if s in raster else None)
        try:
            keylist.remove(s)
        except ValueError:
            pass
    if len(keylist) < 4:
        s = ""
        for c in keylist:
            s.append(c+"\n\t")
        print("The following channels were not mapped: \n"+s)
    image = bytearray.fromhex((defaultB+defaultG+defaultR+defaultA)*sizeX*sizeY)
    for i in range(4):#Only go through channels that are present
        if color[i] is not None:
            for pos in range(sizeX*sizeY):
                image[i+(4*pos)] = int(color[i][pos])
    print(image)
    #Developer area (not implemented yet)
    developerarea = bytearray()
    #Extension area
    extensionarea = bytearray()
    #Footer
    footer = bytearray(int(0).to_bytes(3, END) if extension is None else int(len(header)+len(image)+len(developerarea)).to_bytes(3, END))#Extension area offset
    footer += bytearray(int(0).to_bytes(3, END) if developer is None else int(len(header)+len(image)).to_bytes(3, END))#Extension area offset
    footer += bytearray("TRUEVISION-XFILE.", "ASCII")
    footer += bytearray(int(0).to_bytes(2, END))
    buffer = header+image+developerarea+extensionarea+footer
    if out: print(buffer)
    return buffer


Values.py
#Python 3.2.x only, mofos
import decimal, math, types
from decimal    import Decimal      as D
from math       import ceil, floor, fabs, modf, cos, sin, tan

#Meant to hold brushes that contain values, colors, etc.
class Number:
    settings = {}
    description = "A value that holds a float"
    def man(cls):
        return " \"value\": A floating point number"
    def draw(cls, info, posX, posY):
        return dict.fromkeys(info.colorspace, cls.settings["value"])
    def __init__(cls, settings):
        try:
            cls.settings["value"] = float(settings["value"])
        except:
            print("WTF is this shit?")
            raise ValueError("Sombody obviously doesn't know what a number is.")

class Color:
    settings = {}
    _cache = {}
    description = "A value that holds a float for individual channels (use the None object as a key for a default value)"
    def man(cls):
        return " Channel[*channels]: A floating point number to be used for that particular channel"
    def draw(cls, info, posX, posY):#Should index be an iterator?
        if "input" in cls.settings:
            try:
                color = cls.settings["input"].draw(info, posX, posY)
            except AttributeError:
                color = {}
                for s in info.colorspace:
                    if type(cls.settings["input"][s]) != float and type(cls.settings["input"][s]) != D:
                        raise TypeError("The imbecile that made this thought that values don't need to be a floating point number")
                    if s in cls.settings["input"]:
                        color[s] = cls.settings["input"][s]
                    else:
                        color[s] = cls.setting["input"][None]
        for s in info.colorspace:
            if s in cls.settings:
                color[s] = settings["Channels"][s]
            else:
                color[s] = settings[None]
        return color
    def __init__(cls, settings):
        cls.settings = settings

Brushes.py
#Python 3.2.x only, mofos
import decimal, math, types, Base
from decimal    import Decimal      as D
from math       import ceil, floor, fabs, modf, cos, sin, tan

#Meant to hold brushes that create things (shapes, etc.)
class Circle:
    settings = {}
    _cache = {}
    _description = "Draws a color (or stroke) inside a circle, and another outside of it"
    def __init__(cls, settings):
        cls.settings = settings
        cls._cache["r**2"] = settings["radius"]**2
        for s in ("inside", "outside", "border"):
            if s+"color" in settings:
                if hasattr(settings[s+"color"], "draw"):
                    cls._cache[s+"isstroke"] = True
                elif isinstance(settings[s+"color"], dict):
                    cls._cache[s+"isstroke"] = False
                else:
                    raise TypeError("I don't know WTF you tried to pass as "+s+"color, but it ain't gonna work")
    def man(cls):
        return " \"radius\": The radius of the circle\n \"x\": The center along the X axis\n \"y\": The center along the Y axis\n*\"insidecolor\": The color (or stroke) to paint inside the circle\n*\"outsidecolor\": The color (or stroke) to paint outside the circle"
    def draw(cls, info, posX, posY):
        color = {}
        distance = fabs((posX-D(cls.settings["x"]))**2 + (posY-D(cls.settings["y"]))**2)
        s = "in" if distance <= cls._cache["r**2"] else "out"
        try:
            color = cls.settings[s+"sidecolor"].draw(info, posX, posY) if cls._cache[s+"sideisstroke"] else cls.settings[s+"sidecolor"]
        except AttributeError:#Fix this shit!
            color = cls.settings[s+"sidecolor"]
        if "borderthickness" in cls.settings:#Fix this shit!
            if distance - cls._cache["r**2"] >= cls.settings["borderthickness"]:
                color = cls.settings["bordercolor"].draw(info, posX, posY) if cls._cache["borderisstroke"] else cls.settings["bordercolor"]
        return color

Base.py
#Version 0.006 - Python 3.2.x only, mofos
import decimal, math, copy, os
from decimal    import Decimal as D
from math       import ceil, floor, fabs, modf, cos, sin, tan
LINEINFO = False
PADDING = None#Either an integer or None

class _Info:
    """A class that is used to pass info from one brush to another"""
    colorspace = ""
    resoX, resoY = D("0.00"), D("0.00")#Meant to be updated for zooming, etc.
    def __init__(cls, resoX, resoY, colorspace=""):
        cls.colorspace, cls.resoX, cls.resoY  = colorspace, resoX, resoY

class Texture:
    """A class that holds data about how a texture should render"""
    _sd = {}
    _channels = {}#bytearrays
    input = None
    strokes = None
    def __init__(cls, sizeX, sizeY, dimensionX, dimensionY):
        if sizeX < 1 or sizeY < 1:
            raise ValueError("The person that made this texture (or tried to) is retarded.\n\n Also sizeX or sizeY is less than 1.")
        cls._sd = {"sizeX":int(sizeX), "sizeY":int(sizeY), "dimensionX":(D(dimensionX[0]), D(dimensionX[1])), "dimensionY":(D(dimensionY[0]), D(dimensionY[1])), "resoX":D((D(dimensionX[1])-D(dimensionX[0]))/D(str(sizeX-1))), "resoY":D((D(dimensionY[1])-D(dimensionY[0]))/D(str(sizeY-1)))}
    def render(cls, sizeX=None, sizeY=None, dimensionX=None, dimensionY=None, colorspace="RGBA", bpc=8, thread=(0, 1), altch=None, iterY=None, iterX=None):
        """Asks the input attribute to draw for every sample"""
        print("Thread "+str(thread[0])+" (process "+str(os.getpid()).zfill(4)+") reporting for duty!")
        if cls.input is None: raise TypeError("Some dumbass thinks that you can make an image without painting a stroke. What is this, modern art?")
        if bpc > 8: raise ValueError("You can't have more than one byte per channel (mostly because I can't figure out how to do it without fucking everything up; I would gladly let you do this if I could)")
        maxvalue = (2**bpc)-1
        padto = len(str(sizeY)) if PADDING is None else PADDING
        q = {"sizeX":sizeX, "sizeY":sizeY, "dimensionX":dimensionX, "dimensionY":dimensionY,#This needs to be seperate:
             "resoX":D((D(dimensionX[1])-D(dimensionX[0]))/D(str(sizeX-1))) if (dimensionX is not None and sizeX is not None) else None, "resoY":D((D(dimensionY[1])-D(dimensionY[0]))/D(str(sizeY-1))) if (dimensionY is not None and sizeY is not None) else None}
        sd = cls._sd#Figure out how to merge with line 34?
        sd.update(dict((k, v) for k, v in q.items() if v is not None))
        if hasattr(cls.input, "draw"):
            isstroke = True
        elif isinstance(cls.input, dict):
            isstroke = False
        else:
            raise TypeError("I don't know what the hell you just tried to pass as input, but it ain't gonna work")
        for s in "XY":
            sd["dimension"+s] = (D(sd["dimension"+s][0]), D(sd["dimension"+s][1]))
            sd["reso"+s] = D((D(sd["dimension"+s][1])-D(sd["dimension"+s][0]))/D(str(sd["size"+s]-1)))#Is this needed?
        info = _Info(sd["resoX"], sd["resoY"], colorspace)
        channels = cls._channels if altch is None else altch
        if altch is None:
            b = bytearray.fromhex("00"*sd["sizeX"]*sd["sizeY"])#I'm sure this can be faster
            for s in colorspace:
                cls._channels[s] = copy.copy(b)
        for iY in range(sd["sizeY"]-thread[0]-1, -1, -thread[1]) if iterY is None else iterY:#Figure out how to make it render algebraically (bottom up)
            fY = sd["dimensionY"][1]-(iY*sd["resoY"])
            for iX in (range(sd["sizeX"]) if iterX is None else iterX):
                fX = sd["dimensionX"][1]-(iX*sd["resoX"])#Only gets used once, is it necessary to make it a variable?
                color = cls.input.draw(info, fX, fY) if isstroke else cls.input
                for s in colorspace: channels[s][(iY*sd["sizeY"])+iX] = ceil((color[s]*maxvalue)-.5)
                if iX == sd["sizeX"]-1 and LINEINFO is True:
                    print(str(os.getpid()).zfill(4)+": "+str(iY).zfill(padto)+"@"+("" if fY < 0 else " ")+str(fY)+" cm Y")
        return (cls._channels if thread[1] == 1 else None)
    def draw(cls, info, posX, posY):
        if cls._channels == {}:#Make sure it's already rendered
            cls.render()
        if (posX <= cls._sd["dimensionX"][0] and posX >= cls._sd["dimensionX"][1]) or (posY <= cls._sd["dimensionY"][0] and posY >= cls._sd["dimensionY"][1]):
            return {}
        else:
            pass#Figure out which sample is needed and return it.
    def clearBuffer(cls):
        """Clears the channel buffers"""
        cls._channels = {}
    def getBuffer(cls):
        """Returns the channel buffers"""
        return copy.copy(cls._channels)
